tensorflow/cc/gradients/resource_variable_grad_test.cc

```diff
@@ -0,0 +1,66 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/cc/framework/grad_op_registry.h"
+#include "tensorflow/cc/framework/gradient_checker.h"
+#include "tensorflow/cc/framework/testutil.h"
+#include "tensorflow/cc/gradients/grad_testutil.h"
+#include "tensorflow/core/framework/tensor_testutil.h"
+#include "tensorflow/core/lib/core/status_test_util.h"
+
+namespace tensorflow {
+namespace {
+
+using namespace ops;  // NOLINT(build/namespaces)
+using ops::internal::MirrorPadGrad;
+
+class ResourceVariableGradTest : public ::testing::Test {
+ protected:
+  ResourceVariableGradTest() : scope_(Scope::NewRootScope()) {}
+
+  void RunTest(const Output& x, const TensorShape& x_shape, const Output& y,
+               const TensorShape& y_shape) {
+    TF_ASSERT_OK(scope_.status());
+    float max_error;
+    TF_ASSERT_OK((ComputeGradientError<float, float, float>(
+        scope_, {x}, {x_shape}, {y}, {y_shape}, &max_error)));
+    EXPECT_LT(max_error, 1e-3);
+  }
+
+  void RunTest(const OutputList& xs, const std::vector<TensorShape>& x_shapes,
+               const OutputList& ys, const std::vector<TensorShape>& y_shapes) {
+    TF_ASSERT_OK(scope_.status());
+    float max_error;
+    TF_ASSERT_OK((ComputeGradientError<float, float, float>(
+        scope_, xs, x_shapes, ys, y_shapes, &max_error)));
+    EXPECT_LT(max_error, 1e-3);
+  }
+
+  Scope scope_;
+};
+
+TEST_F(ResourceVariableGradTest, ReadVariableOpGrad) {
+  TensorShape shape({5, 2});
+  auto x = Placeholder(scope_, DT_FLOAT, Placeholder::Shape(shape));
+  auto var = VarHandleOp(scope_, DT_FLOAT, {});
+  auto init = AssignVariableOp(scope_, var, x);
```

Comment:
Agreed, I'm not (intentionally) testing that.  I'm attempting to work with the existing test harness which afaik requires me to use a placeholder as input, ideally I'd get a variable and read from it.  The assignment is the equivalent of `tf.Variable(x)` so that it's initialized.  

The test report 404s now, but if I remember right this test passed.  I'll fix the python side and see.

Reply:
I see. Yeah, I guess if you want to use the numeric checker you need to do something like this.

What I don't understand is why there is a gradient of `x` with respect to `y`; `var` with respect to `y` makes sense to me, but if assignment has a gradient too then there's a rabbit hole that I don't think is implemented correctly right now (the gradient only flowing to the second assignment at execution time, which isn't statically known).

Reply:
I'd just manually take a gradient of the resource handle with respect to the output of the ReadVariableOp (or with respect to the Sum to reduce it to a scalar if it isn't already). It's just a ones tensor with the shape of the ReadVariableOp output, so you don't really need the gradient checker. (Or if you want to make it more interesting, you can do something after the read op; either way the read op is identity in the backward pass.)

At least, that's how the tape models variables in Python. There is a can of worms around the gradient of a scalar-shape resource handle being a non-scalar (shape/dtype of the ReadVariableOp output). There's a question of how much we want to do for these graph-only C++ gradients vs. doing it once with the eager/graph agnostic API.

Reply:
I didn't run into the scalar shape issue.  I don't think we need to do much with the C++ API, I just want this as a bandaid so I can start working on variables in the Java API.